chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/onetime/conformancemethodconversion/ConformanceMethodConversionJob.java

```diff
@@ -0,0 +1,287 @@
+import org.springframework.transaction.support.TransactionTemplate;
+import org.springframework.web.context.support.SpringBeanAutowiringSupport;
+
+import gov.healthit.chpl.FeatureList;
+import gov.healthit.chpl.auth.user.User;
+import gov.healthit.chpl.conformanceMethod.entity.ConformanceMethodEntity;
+import gov.healthit.chpl.dao.impl.BaseDAOImpl;
+import gov.healthit.chpl.domain.CertificationCriterion;
+import gov.healthit.chpl.domain.CertificationResult;
+import gov.healthit.chpl.domain.CertificationResultTestProcedure;
+import gov.healthit.chpl.domain.CertifiedProductSearchDetails;
+import gov.healthit.chpl.entity.listing.CertificationResultConformanceMethodEntity;
+import gov.healthit.chpl.entity.listing.CertificationResultTestProcedureEntity;
+import gov.healthit.chpl.scheduler.job.CertifiedProduct2015Gatherer;
+import gov.healthit.chpl.service.CertificationCriterionService;
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.Setter;
+import lombok.extern.log4j.Log4j2;
+
+@Log4j2(topic = "conformanceMethodConversionJobLogger")
+public class ConformanceMethodConversionJob extends CertifiedProduct2015Gatherer implements Job {
+    private static final String WILDCARD = "*";
+
+    @Autowired
+    private CertificationCriterionService certificationCriterionService;
+
+    @Autowired
+    private ConformanceMethodConversionDAO conformanceMethodConversionDAO;
+
+    @Autowired
+    private JpaTransactionManager txManager;
+
+    @Autowired
+    private FF4j ff4j;
+
+    private List<ConversionRule> conversionRules;
+
+    @Override
+    public void execute(JobExecutionContext context) throws JobExecutionException {
+        SpringBeanAutowiringSupport.processInjectionBasedOnCurrentContext(this);
+        LOGGER.info("********* Starting the Conformance Method Conversion job. *********");
+
+        if (ff4j.check(FeatureList.CONFORMANCE_METHOD)) {
+            conversionRules = getConversionRules();
+
+            // We need to manually create a transaction in this case because of how AOP works. When a method is
+            // annotated with @Transactional, the transaction wrapper is only added if the object's proxy is called.
+            // The object's proxy is not called when the method is called from within this class. The object's proxy
+            // is called when the method is public and is called from a different object.
+            // https://stackoverflow.com/questions/3037006/starting-new-transaction-in-spring-bean
+            TransactionTemplate txTemplate = new TransactionTemplate(txManager);
+            txTemplate.setPropagationBehavior(TransactionDefinition.PROPAGATION_REQUIRES_NEW);
+            txTemplate.execute(new TransactionCallbackWithoutResult() {
+
+                @Override
+                protected void doInTransactionWithoutResult(TransactionStatus status) {
+                    try {
+
+                        getAll2015CertifiedProducts(LOGGER, 2).stream()
```

Comment:
I have some concern extending the CertifiedProduct2015Gatherer and using this method which retrieves the 2015 listing details from the cache. The concern is due to some issues I ran into with the merge and split of developers not recording activity (the jobs were getting the "before" listing and "after" listing from the cache, so not registering the change). If this job were guaranteed to only run immediately following the cache being populated by the overnight jobs that would alleviate my concern. If it is going to be run at any other time then I am not sure about using the cached listing details since the test procedures for that listing could potentially have been changed between cache population and job run. I completely realize this is unlikely but the cached data burned me twice for that recent ticket so I am sensitive to it.